### PR TITLE
Carry 10864: ANSI terminal emulation for windows

### PR DESCRIPTION
**Pull request by tiborvass**
Closes #10864 

This is the `win-tty-no-chan` branch from @sachin-jayant-joshi. It contains a couple of fixes.
Let's get this merged, we can always add more fixes to it.

### CODE REVIEW

File: pkg/term/winconsole/console_windows.go [hunk 1]
```diff
@@ -0,0 +1,1124 @@
...
+		panic("This should never happen as it is predefined handle.")
+	}
+	screenBufferInfo, err := GetConsoleScreenBufferInfo(uintptr(handle))
+	if err == nil {
+		handler.screenBufferInfo = screenBufferInfo
+	}
+
+	// Set the window size
+	SetWindowSize(uintptr(handle), DEFAULT_WIDTH, DEFAULT_HEIGHT, DEFAULT_HEIGHT)
+	buffer = make([]CHAR_INFO, screenBufferInfo.MaximumWindowSize.X*screenBufferInfo.MaximumWindowSize.Y)
```
Comment (tiborvass) [inline pkg/term/winconsole/console_windows.go]: this panics on jenkins windows, my guess is that we don't check for the error returned by GetConsoleScreenBufferInfo

File: pkg/term/winconsole/console_windows.go [hunk 2]
```diff
@@ -0,0 +1,1124 @@
...
+func GetConsoleScreenBufferInfo(fileDesc uintptr) (*CONSOLE_SCREEN_BUFFER_INFO, error) {
+	var info CONSOLE_SCREEN_BUFFER_INFO
+	r, _, err := getConsoleScreenBufferInfoProc.Call(uintptr(fileDesc), uintptr(unsafe.Pointer(&info)), 0)
+	if r == 0 {
+		if err != nil {
+			return nil, err
+		}
+		return nil, syscall.EINVAL
+	}
+	return &info, nil
```
Comment (tiborvass) [inline pkg/term/winconsole/console_windows.go]: this feels wrong, if `r != 0`, according to MSDN docs, it means it's an error. However, we return no error.

**Comment by jfrazelle** [comment]
working pretty well for me, LGTM

**Comment by icecrime** [comment]
Still got a few quirks (see for example below with `htop`), but overall super cool :+1: Thanks @sachin-jayant-joshi @ahmetalpbalkan @tiborvass!

screen shot 2015-03-22 at 18 15 47

After first update, some of the text on the bottom (which btw should probably sit at the very bottom) get overridden:

screen shot 2015-03-22 at 18 15 49

**Comment by icecrime** [comment]
LGTM

**Comment by ahmetalpbalkan** [comment]
wow you guys are quick. :smile:  @sachin-jayant-joshi was just showing me the commit he sent to `win-tty-no-chan` and he was gonna push that to PR soon. :smiley: anyway got merged. that's cool.

**Comment by jfrazelle** [comment]
do we need that too? he can open another PR :)

**Comment by ahmetalpbalkan** [comment]
Thats the one tinor cherrypicked here. :-)

**Comment by jfrazelle** [comment]
jk my bad Sunday's lol

**Comment by sachin-jayant-joshi** [comment]
@tiborvass @icecrime @ahmetalpbalkan
Thanks a lot for all the help!!
Really appreciate it.
We'll continue fixing bugs..

**Comment by tiborvass** [comment]
Hey @sachin-jayant-joshi please review my PR even though it's merged. If
you see anything weird let me know. I did some refactor and style cleanup.

**Pull request merged**